### PR TITLE
Handle fabric removal in the Door Lock Cluster and fix GetUser/GetCredentialStatus commands

### LINKED ISSUE #14919: Door-Lock-Server returning Next Available Index on GetUserCommand and GetCredentialStatusCommand instead of next Occupied Index
#### Problem
Door-Lock-Server.cpp is returning the Next Available user index on Get User command
Door-Lock-Server.cpp is returning the Next Available credential index on Get Credential Status command

per spec, the NextIndex for both of these commands should instead return the next *occupied* index.

### PR DESCRIPTION
**Pull request by Morozov-5F**
#### Problem
* Door lock cluster did not handle the fabric removal properly.
* Fixes #14919.

#### Change overview
* Now door lock cluster fixes the fabric removal according to spec: 
  * If the fabric was deleted, the creator/modifier fabric index in user/credential records is set to 0 (both supported in SDK and Linux App)
  * If the credential was created outside Matter IM the creator/modifier fabric index in user/credential records is set to null (only SDK support as there's no way to create users/credentials outside Matter IM for now in the Linux App).
* GetUserResponse/GetCredentialStatusResponse status command now return next *occupied* index in the payload.

#### Testing
* Ran DL_UsersCredentials, DL_LockUnlock and DL_Schedules YAML tests.
* To test the fabric removal I've used the following procedure:
  1. Run the Linux `chip-lock-app`
  2. Pair with the app using the `chip-tool`:
     ```
     ? ./chip-tool pairing onnetwork 1 20202021
     ```
  3. Open commissioning window again so we can pair with another fabric:
     ```
     ? ./chip-tool pairing open-commissioning-window 1 0 200 1000 3840
     ```
  4. Pair with different fabric:
     ```
     ? ./chip-tool pairing onnetwork 1 20202021 --commissioner-name beta
     ```
  5. Create new credentials and users from the first fabric: 
     ```
     ? ./chip-tool doorlock set-credential 0 '{ "credentialType" : 1 , "credentialIndex" : 1 }' 123456 null null null 1 1 --timedInteractionTimeoutMs 1000
     ```
  6. Create new credentials and users from the second fabric:
     ```
     ? ./chip-tool doorlock set-credential 0 '{ "credentialType" : 2 , "credentialIndex" : 2 }' 12345678901213 null null null 1 1 --timedInteractionTimeoutMs 1000 --commissioner-name beta
     ```
  7. Delete the second fabric:
     ```
     ? ./chip-tool operationalcredentials remove-fabric 2 1 0 --commissioner-name beta
     ```
  8. Make sure the fabric index in the users/credentials created by the first fabric is the same:
     ```
     ? chip-tool doorlock get-user 1 1 1
     ? chip-tool doorlock get-credential-status '{ "credentialType" : 1 , "credentialIndex" : 1 }' 1 1
     ```
  9. Make sure the fabric index in the users/credentials created by the second fabric is set to `0`:
     ```
     ? chip-tool doorlock get-user 2 1 1
     ? chip-tool doorlock get-credential-status '{ "credentialType" : 2 , "credentialIndex" : 2 }' 1 1
     ```

### CODE REVIEW

File: examples/lock-app/efr32/src/LockManager.cpp [hunk 1]
```diff
@@ -318,6 +322,10 @@ bool LockManager::GetCredential(chip::EndpointId endpointId, uint16_t credential
     }
     credential.credentialType = credentialInStorage.credentialType;
     credential.credentialData = credentialInStorage.credentialData;
+    // So far there's no way to actually create the credential outside the matter, so here we always set the creation/modification
```
Comment (bzbarsky-apple) [inline examples/lock-app/efr32/src/LockManager.cpp:325]:
```suggestion
    // So far there's no way to actually create the credential outside Matter, so here we always set the creation/modification
```

File: src/app/clusters/door-lock-server/door-lock-server.cpp [hunk 2]
```diff
@@ -655,54 +687,34 @@ void DoorLockServer::GetCredentialStatusCommandHandler(
...
-    SuccessOrExit(err = commandObj->PrepareCommand(path));
-    VerifyOrExit((writer = commandObj->GetCommandDataIBTLVWriter()) != nullptr, err = CHIP_ERROR_INCORRECT_STATE);
-    SuccessOrExit(err = writer->Put(TLV::ContextTag(to_underlying(ResponseFields::kCredentialExists)), credentialExists));
+    Commands::GetCredentialStatusResponse::Type response{ .credentialExists = credentialExists };
     if (credentialExists)
     {
         if (0 != userIndexWithCredential)
         {
-            SuccessOrExit(err = writer->Put(TLV::ContextTag(to_underlying(ResponseFields::kUserIndex)), userIndexWithCredential));
+            response.userIndex = Nullable<uint16_t>(userIndexWithCredential);
```
Comment (bzbarsky-apple) [inline src/app/clusters/door-lock-server/door-lock-server.cpp:695]:
```suggestion
            response.userIndex.SetNonNull(userIndexWithCredential);
```

File: src/app/clusters/door-lock-server/door-lock-server.cpp [hunk 3]
```diff
@@ -655,54 +687,34 @@ void DoorLockServer::GetCredentialStatusCommandHandler(
...
         }
-        if (kUndefinedFabricIndex != credentialInfo.createdBy)
+        if (credentialInfo.creationSource == DlAssetSource::kMatterIM)
         {
-            SuccessOrExit(
-                err = writer->Put(TLV::ContextTag(to_underlying(ResponseFields::kCreatorFabricIndex)), credentialInfo.createdBy));
+            response.creatorFabricIndex = Nullable<chip::FabricIndex>(credentialInfo.createdBy);
```
Comment (bzbarsky-apple) [inline src/app/clusters/door-lock-server/door-lock-server.cpp:699]:
```suggestion
            response.creatorFabricIndex.SetNonNull(credentialInfo.createdBy);
```

File: src/app/clusters/door-lock-server/door-lock-server.cpp [hunk 4]
```diff
@@ -655,54 +687,34 @@ void DoorLockServer::GetCredentialStatusCommandHandler(
...
         }
-        if (kUndefinedFabricIndex != credentialInfo.lastModifiedBy)
+        if (credentialInfo.modificationSource == DlAssetSource::kMatterIM)
         {
-            SuccessOrExit(err = writer->Put(TLV::ContextTag(to_underlying(ResponseFields::kLastModifiedFabricIndex)),
-                                            credentialInfo.lastModifiedBy));
+            response.lastModifiedFabricIndex = Nullable<chip::FabricIndex>(credentialInfo.lastModifiedBy);
```
Comment (bzbarsky-apple) [inline src/app/clusters/door-lock-server/door-lock-server.cpp:703]:
```suggestion
            response.lastModifiedFabricIndex.SetNonNull(credentialInfo.lastModifiedBy);
```

File: src/app/clusters/door-lock-server/door-lock-server.cpp [hunk 5]
```diff
@@ -655,54 +687,34 @@ void DoorLockServer::GetCredentialStatusCommandHandler(
...
         }
     }
-    if (findUnoccupiedCredentialSlot(commandPath.mEndpointId, credentialType, static_cast<uint16_t>(credentialIndex + 1),
-                                     nextCredentialIndex))
+    uint16_t nextCredentialIndex = 0;
+    if (findOccupiedCredentialSlot(commandPath.mEndpointId, credentialType, static_cast<uint16_t>(credentialIndex + 1),
+                                   nextCredentialIndex))
     {
-        SuccessOrExit(err = writer->Put(TLV::ContextTag(to_underlying(ResponseFields::kNextCredentialIndex)), nextCredentialIndex));
+        response.nextCredentialIndex = Nullable<uint16_t>(nextCredentialIndex);
```
Comment (bzbarsky-apple) [inline src/app/clusters/door-lock-server/door-lock-server.cpp:710]:
```suggestion
        response.nextCredentialIndex.SetNonNull(nextCredentialIndex);
```

File: src/app/clusters/door-lock-server/door-lock-server.cpp [hunk 6]
```diff
@@ -1204,6 +1216,30 @@ bool DoorLockServer::HasFeature(chip::EndpointId endpointId, DoorLockFeature fea
...
 }
 
+bool DoorLockServer::OnFabricRemoved(chip::EndpointId endpointId, chip::FabricIndex fabricIndex)
+{
+    emberAfDoorLockClusterPrintln(
+        "[OnFabricRemoved] Handling a fabric removal from the door lock server [endpointId=%d,fabricIndex=%d]", endpointId,
+        fabricIndex);
+
+    // Iterate over all the users and clean up the deleted fabric
+    if (!cleanFabricFromUsers(endpointId, fabricIndex))
```
Comment (bzbarsky-apple) [inline src/app/clusters/door-lock-server/door-lock-server.cpp:1226]:
```suggestion
    if (!clearFabricFromUsers(endpointId, fabricIndex))
```

File: src/app/clusters/door-lock-server/door-lock-server.cpp [hunk 7]
```diff
@@ -1384,6 +1446,42 @@ bool DoorLockServer::findUnoccupiedUserSlot(chip::EndpointId endpointId, uint16_
...
+bool DoorLockServer::findOccupiedCredentialSlot(chip::EndpointId endpointId, DlCredentialType credentialType, uint16_t startIndex,
+                                                uint16_t & credentialIndex)
+{
+    uint16_t maxNumberOfCredentials = 0;
+    if (!getMaxNumberOfCredentials(endpointId, credentialType, maxNumberOfCredentials))
+    {
+        return false;
+    }
+
+    // Programming PIN index starts with 0, and it is assumed that it is unique. Therefor different bounds checking for that
```
Comment (bzbarsky-apple) [inline src/app/clusters/door-lock-server/door-lock-server.cpp:1458]:
```suggestion
    // Programming PIN index starts with 0, and it is assumed that it is unique. Therefore different bounds checking for that
```

File: src/app/clusters/door-lock-server/door-lock-server.cpp [hunk 8]
```diff
@@ -1714,6 +1812,43 @@ EmberAfStatus DoorLockServer::clearUser(chip::EndpointId endpointId, chip::Fabri
...
+            continue;
+        }
+
+        // Filter out unoccupied slots and users that don't have corresponding fabricIndex in the created/modified fields
+        if (DlUserStatus::kAvailable == user.userStatus || (fabricIndex != user.createdBy && fabricIndex != user.lastModifiedBy))
+        {
+            continue;
+        }
+
+        if (!emberAfPluginDoorLockSetUser(endpointId, userIndex, kUndefinedFabricIndex, kUndefinedFabricIndex, user.userName,
```
Comment (bzbarsky-apple) [inline src/app/clusters/door-lock-server/door-lock-server.cpp:1839]: No, this is wrong.  If the user is created by fabric index 2, then modified by fabric index 3, then 3 is removed, then why would we reset the "Created by" state?

File: src/app/clusters/door-lock-server/door-lock-server.cpp [hunk 9]
```diff
@@ -2578,6 +2713,83 @@ EmberAfStatus DoorLockServer::clearCredentials(chip::EndpointId endpointId, chip
...
+            continue;
+        }
+
+        if (DlCredentialStatus::kAvailable == credential.status ||
+            (credential.createdBy != fabricToRemove && credential.lastModifiedBy != fabricToRemove))
+        {
+            continue;
+        }
+
+        if (!emberAfPluginDoorLockSetCredential(endpointId, credentialIndex, kUndefinedFabricIndex, kUndefinedFabricIndex,
```
Comment (bzbarsky-apple) [inline src/app/clusters/door-lock-server/door-lock-server.cpp:2751]: Again, this needs to reset only the fields that match fabricToRemove....

File: src/app/clusters/door-lock-server/door-lock-server.cpp [hunk 10]
```diff
@@ -2578,6 +2713,83 @@ EmberAfStatus DoorLockServer::clearCredentials(chip::EndpointId endpointId, chip
...
+            continue;
+        }
+    }
+
+    return true;
+}
+
+bool DoorLockServer::clearFabricFromCredentials(chip::EndpointId endpointId, chip::FabricIndex fabricToRemove)
+{
+    if (SupportsPFID(endpointId))
```
Comment (bzbarsky-apple) [inline src/app/clusters/door-lock-server/door-lock-server.cpp:2767]: What is PFID?

File: src/app/clusters/door-lock-server/door-lock-server.h [hunk 11]
```diff
@@ -368,15 +379,25 @@ enum class DlCredentialStatus : uint8_t
...
-    DlCredentialStatus status;        /**< Indicates if credential slot is occupied or not. */
-    DlCredentialType credentialType;  /**< Specifies the type of the credential (PIN, RFID, etc.). */
-    chip::ByteSpan credentialData;    /**< Credential data bytes. */
-    chip::FabricIndex createdBy;      /**< ID of the fabric that created the user. */
+    DlCredentialStatus status;       /**< Indicates if credential slot is occupied or not. */
+    DlCredentialType credentialType; /**< Specifies the type of the credential (PIN, RFID, etc.). */
+    chip::ByteSpan credentialData;   /**< Credential data bytes. */
+
+    DlAssetSource creationSource;
+    chip::FabricIndex createdBy; /**< ID of the fabric that created the user. */
```
Comment (bzbarsky-apple) [inline src/app/clusters/door-lock-server/door-lock-server.h:398]:
```suggestion
    chip::FabricIndex createdBy; /**< Index of the fabric that created the user. */
```
and similar for the other members.

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]

File: src/app/clusters/door-lock-server/door-lock-server.cpp (hunk 8 above)
Comment (Morozov-5F) [reply src/app/clusters/door-lock-server/door-lock-server.cpp:1839]: Good catch, thanks

File: src/app/clusters/door-lock-server/door-lock-server.cpp (hunk 10 above)
Comment (Morozov-5F) [reply src/app/clusters/door-lock-server/door-lock-server.cpp:2767]: That stands for "I can't English" :) My bad

**Review by bzbarsky-apple — APPROVED** [review]

File: src/app/clusters/door-lock-server/door-lock-server.cpp [hunk 12]
```diff
@@ -1204,6 +1216,30 @@ bool DoorLockServer::HasFeature(chip::EndpointId endpointId, DoorLockFeature fea
     return success && ((featureMap & to_underlying(feature)) != 0);
 }
 
+bool DoorLockServer::OnFabricRemoved(chip::EndpointId endpointId, chip::FabricIndex fabricIndex)
```
[the remark this answers is not in the document]
Comment (Morozov-5F) [reply src/app/clusters/door-lock-server/door-lock-server.cpp:1219]: That might be a good idea, will do

File: src/app/clusters/door-lock-server/door-lock-server.cpp [hunk 13]
```diff
@@ -2578,6 +2723,93 @@ EmberAfStatus DoorLockServer::clearCredentials(chip::EndpointId endpointId, chip
...
+    if (!getMaxNumberOfCredentials(endpointId, credentialType, maxNumberOfCredentials))
+    {
+        ChipLogError(
+            Zcl,
+            "[clearFabricFromCredentials] Unable to get max number of credentials to clear - can't get max number of credentials "
+            "[endpointId=%d,credentialType=%u]",
+            endpointId, to_underlying(credentialType));
+        return false;
+    }
+
```
[the remark this answers is not in the document]
Comment (Morozov-5F) [reply src/app/clusters/door-lock-server/door-lock-server.cpp:2739]: Great catch, thanks!

File: src/app/clusters/door-lock-server/door-lock-server.cpp [hunk 14]
```diff
@@ -2578,6 +2723,93 @@ EmberAfStatus DoorLockServer::clearCredentials(chip::EndpointId endpointId, chip
     return EMBER_ZCL_STATUS_SUCCESS;
 }
 
+bool DoorLockServer::clearFabricFromCredentials(chip::EndpointId endpointId, DlCredentialType credentialType,
```
[the remark this answers is not in the document]
Comment (Morozov-5F) [reply src/app/clusters/door-lock-server/door-lock-server.cpp:2726]: Yeah, makes sense

File: examples/lock-app/efr32/src/LockManager.cpp [hunk 15]
```diff
@@ -318,6 +322,10 @@ bool LockManager::GetCredential(chip::EndpointId endpointId, uint16_t credential
     }
     credential.credentialType = credentialInStorage.credentialType;
     credential.credentialData = credentialInStorage.credentialData;
+    // So far there's no way to actually create the credential outside Matter, so here we always set the creation/modification
+    // source to Matter
+    credential.creationSource     = DlAssetSource::kMatterIM;
+    credential.modificationSource = DlAssetSource::kMatterIM;
```
[the remark this answers is not in the document]
Comment (Morozov-5F) [reply examples/lock-app/efr32/src/LockManager.cpp:328]: I actually forgot about the efr32 app, should be fixed now, thanks.

File: src/app/clusters/door-lock-server/door-lock-server.cpp [hunk 16]
```diff
@@ -1242,6 +1242,7 @@ bool DoorLockServer::OnFabricRemoved(chip::EndpointId endpointId, chip::FabricIn
     {
         ChipLogError(Zcl, "[OnFabricRemoved] Unable to cleanup fabric from users - internal error [endpointId=%d,fabricIndex=%d]",
                      endpointId, fabricIndex);
+        return false;
```
Comment (tehampson) [inline src/app/clusters/door-lock-server/door-lock-server.cpp:1245]: Nit comment. I am unfamiliar with this overall, so feel free to ignore and resolve this comment if you don't think it applies or makes sense

In here, and the other methods `clearFabricFromUsers`, and `clearFabricFromCredentials`, does it make sense to start of with `bool returnValue = true` at the top and set it to false in these if statements and just `return returnValue` so we at least attempt to clear other things even if we fail on one thing?

**Review by tehampson — APPROVED** [review]

File: src/app/clusters/door-lock-server/door-lock-server.cpp (hunk 16 above)
Comment (Morozov-5F) [reply src/app/clusters/door-lock-server/door-lock-server.cpp:1245]: Thanks, good idea! I'm just unsure if it actually makes sense to keep going if calls to the app fail. We just need to choose the strategy and apply it everywhere. I am currently not ready to make this decision, maybe it makes sense to create an issue for this.

Comment (tehampson) [reply src/app/clusters/door-lock-server/door-lock-server.cpp:1245]: Sounds good to me. You can resolve this comment and merge the PR, and create the issue

**Review by woody-apple — APPROVED** [review]

**Pull request merged**